test/unit/attachment_data_visibility_test.rb

```diff
@@ -0,0 +1,643 @@
+          attributes[:to_replace_id] = attachment_data.id
+          attachment.update_attributes!(attachment_data_attributes: attributes)
+        end
+
+        it 'is not deleted' do
+          refute attachment_data.reload.deleted?
+        end
+
+        it 'is draft' do
+          assert attachment_data.reload.draft?
+        end
+
+        it 'is replaced' do
+          assert attachment_data.reload.replaced?
+        end
+
+        context 'when edition is published' do
+          before do
+            edition.major_change_published_at = Time.zone.now
+            edition.force_publish!
+          end
+
+          it 'is not deleted' do
+            refute attachment_data.reload.deleted?
+          end
+
+          it 'is draft' do
+            assert attachment_data.reload.draft?
+          end
+
+          it 'is replaced' do
+            assert attachment_data.reload.replaced?
+          end
+        end
+      end
+
+      context 'when edition is published' do
+        before do
+          edition.major_change_published_at = Time.zone.now
+          edition.force_publish!
+        end
+
+        it 'is not deleted' do
+          refute attachment_data.reload.deleted?
+        end
+
+        it 'is not draft' do
+          refute attachment_data.reload.draft?
+        end
+
+        it 'is not unpublished' do
+          refute attachment_data.reload.unpublished?
+        end
+
+        it 'has no unpublished edition' do
+          assert_nil attachment_data.reload.unpublished_edition
+        end
+
+        context 'and new edition is created' do
+          let(:new_edition) { edition.create_draft(user) }
+          let(:new_attachable) { new_edition }
+          let(:new_attachment) { new_attachable.attachments.first }
+
+          before do
+            new_edition.reload
+          end
+
+          it 'is not deleted' do
+            refute attachment_data.reload.deleted?
+          end
+
+          it 'is not draft' do
+            refute attachment_data.reload.draft?
+          end
+
+          it 'is not unpublished' do
+            refute attachment_data.reload.unpublished?
+          end
+
+          it 'has no unpublished edition' do
+            assert_nil attachment_data.reload.unpublished_edition
+          end
+
+          context 'new edition is access-limited' do
+            before do
+              new_edition.change_note = 'change-note'
+              new_edition.access_limited = true
+              new_edition.save!
+            end
+
+            it 'is not accessible to anonymous user' do
+              refute attachment_data.reload.accessible_to?(anonymous_user)
+            end
+
+            it 'is accessible to user in same organisation' do
+              assert attachment_data.reload.accessible_to?(user_in_same_organisation)
+            end
+
+            it 'is accessible to user in another organisation' do
+              assert attachment_data.reload.accessible_to?(user_in_another_organisation)
```

Comment:
It took us a while to work out why this was accessible to a user in a different organisation. It's because the `AttachmentData` has been published and is therefore not draft, which means that it can't/shouldn't be access limited.

We discussed adding some more information to the commit note to explain this.

Reply:
I've added the following paragraph to the relevant commit note:

> Note that if an attachment is added to a draft edition, the edition is published, a new draft edition is created and marked as access-limited, then the attachment is accessible to a user in another organisation, because the published edition cannot be access-limited.

Reply:
It turns out I forgot to push up the changes to the commit note before I merged the PR, so the above comment will have to suffice - sorry!